src/Common/ZooKeeper/ZooKeeperImpl.cpp

```diff
@@ -1078,8 +1138,21 @@ void ZooKeeper::pushRequest(RequestInfo && info)
             }
         }
 
-        if (unlikely(send_inject_fault) && send_inject_fault.value()(thread_local_rng))
-            throw Exception(Error::ZSESSIONEXPIRED, "Session expired (fault injected on send)");
+        if (unlikely(inject_fault))
+        {
+            bool fault_session = false;
+            bool fault_timeout = false;
+            if (auto lock_guard = shared_mutex_try_lock_shared(&inject_mutex))
+            {
+                fault_session = send_session_inject_fault && send_session_inject_fault.value()(thread_local_rng);
+                fault_timeout = send_timeout_inject_fault && send_timeout_inject_fault.value()(thread_local_rng);
```

Comment:
Thanks a lot for the pointers. The errors were likely due to too many requests to ZK and the machine being overloaded (15k replicated tables in a staging server that run both CH, ZK and other services). For example:
```
`clickhouse-server.log.5.gz:2023.03.18 18:14:36.357220 [ 13812 ] {} <Information> ZooKeeperClient: Finalizing session 109789916112879673: finalization_started=false, queue_finished=false, reason=Operation timeout on Exists /clickhouse/tables/01-01/d_2d06a1.t_b2272b73e7cd41f7b94b2518123153a1/replicas/10.154.15.217/is_active`
```

I'll try to simulate that. It seems that it'd make sense to have something like:
* Ability to choose what kind of error code to throw.
* Ability to sleep to simulate timeouts or receive or send by sleeping at proper places (for a fix amount just over the 10s timeout). Using the ThreadFuzzer might make sense, but I don't know how easy it'd be to wrap the `ZooKeeper` functions.
* Ability to avoid faults at start

I'll see what I can do / what's worth doing now.

Reply:
I've started a different approach using sleep and `enable_fault_injections_during_startup`. I still need to do some more tests and verifications, but I hope it's simpler that the other approach.

Reply:
It seems there are some binaries that depend on zookeeper but not on context, so they fail to build as `Context::setServerCompletelyStarted` is not present. I'll change it to set the value from the startup code instead of polling it every time.